symengine/rewrite.cpp

```diff
@@ -130,4 +130,112 @@ RCP<const Basic> rewrite_as_exp(const RCP<const Basic> &x)
     return b.apply(x);
 }
 
+class RewriteAsSin : public BaseVisitor<RewriteAsSin, TransformVisitor>
+{
+public:
+    using TransformVisitor::bvisit;
+
+    RewriteAsSin() : BaseVisitor<RewriteAsSin, TransformVisitor>()
+    {
+    }
+
+    void bvisit(const Cos &x)
+    {
+        auto farg = x.get_arg();
+        auto newarg = apply(farg);
+        result_ = sin(unevaluated_expr(add(newarg, div(pi, integer(2)))));
+    }
+
+    void bvisit(const Tan &x)
+    {
+        auto farg = x.get_arg();
+        auto newarg = apply(farg);
+        result_ = div(mul(integer(2), pow(sin(newarg), integer(2))),
+                      sin(mul(integer(2), newarg)));
+    }
+
+    void bvisit(const Cot &x)
+    {
+        auto farg = x.get_arg();
+        auto newarg = apply(farg);
+        result_ = div(sin(mul(integer(2), newarg)),
+                      mul(integer(2), pow(sin(newarg), integer(2))));
+    }
+
+    void bvisit(const Csc &x)
+    {
+        auto farg = x.get_arg();
+        auto newarg = apply(farg);
+        result_ = div(integer(1), sin(newarg));
+    }
+
+    void bvisit(const Sec &x)
+    {
+        auto farg = x.get_arg();
+        auto newarg = apply(farg);
+        result_ = div(integer(1),
+                      sin(unevaluated_expr(add(newarg, div(pi, integer(2))))));
+    }
+};
+
+RCP<const Basic> rewrite_as_sin(const RCP<const Basic> &x)
+{
+    RewriteAsSin b;
+    return b.apply(x);
+}
+
+class RewriteAsCos : public BaseVisitor<RewriteAsCos, TransformVisitor>
+{
+public:
+    using TransformVisitor::bvisit;
+
+    RewriteAsCos() : BaseVisitor<RewriteAsCos, TransformVisitor>()
+    {
+    }
+
+    void bvisit(const Sin &x)
+    {
+        auto farg = x.get_arg();
+        auto newarg = apply(farg);
+        result_ = cos(unevaluated_expr(add(newarg, div(pi, integer(2)))));
```

Comment:
should this be `sin(x) = cos(x - pi/2)`?
```suggestion
        result_ = cos(unevaluated_expr(sub(newarg, div(pi, integer(2)))));
```

Reply:
Of course. Good catch!